{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a6fc2523-648f-4464-b948-065f17ba1c8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(r\"C:\\Users\\liamm\\Desktop\\chr1.GRCh38.excerpt.fasta\", 'r') as f:\n",
    "        for line in f:\n",
    "            # ignore header line with genome information\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ca6d800e-17f3-4658-9b49-d1b2b9a81138",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readFastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "    with open(filename) as fh:\n",
    "        while True:\n",
    "            fh.readline()  # skip name line\n",
    "            seq = fh.readline().rstrip()  # read base sequence\n",
    "            fh.readline()  # skip placeholder line\n",
    "            qual = fh.readline().rstrip() # base quality line\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "    return sequences, qualities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5622659a-11f9-478f-8b59-17bc34a257bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "def naive(p, t):\n",
    "    occurrences = []\n",
    "    align_count = 0\n",
    "    compare_count = 0\n",
    "    for i in range(len(t) - len(p) + 1):  # loop over alignments\n",
    "        match = True\n",
    "        for j in range(len(p)):# loop over characters\n",
    "            compare_count += 1\n",
    "            if t[i+j] != p[j]:  # compare characters\n",
    "                match = False\n",
    "                break\n",
    "        if match:\n",
    "            occurrences.append(i)  # all chars matched; record\n",
    "        align_count += 1\n",
    "    return occurrences,align_count,compare_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "80fe6a78-da55-4fb4-bc94-51f6527e8a63",
   "metadata": {},
   "outputs": [],
   "source": [
    "s = 'GGCGCGGTGGCTCACGCCTGTAATCCCAGCACTTTGGGAGGCCGAGG'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d11b8a38-60f4-425b-9fe0-362654e79a82",
   "metadata": {},
   "outputs": [],
   "source": [
    "genome = readGenome(r\"C:\\Users\\liamm\\Desktop\\chr1.GRCh38.excerpt.fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "d8e1a6e8-9d01-4e57-a197-cd2703f01946",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TTGAATGCTGAAATCAGCAGGTAATATATGATAATAGAGAAAGCTATCCCGAAGGTGCATAGGTCAACAATACTTGAGCCTAACTCAGTAGATCCTAAAA'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genome[:100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "56039dc4-b626-406b-83c6-3ef4bf3f67ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([56922], 799954, 984143)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "naive(s,genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "d97ec819-b8fc-45fb-8165-1fc3d3e8f8f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "import string\n",
    "\n",
    "def z_array(s):\n",
    "    \"\"\" Use Z algorithm (Gusfield theorem 1.4.1) to preprocess s \"\"\"\n",
    "    assert len(s) > 1\n",
    "    z = [len(s)] + [0] * (len(s)-1)\n",
    "\n",
    "    # Initial comparison of s[1:] with prefix\n",
    "    for i in range(1, len(s)):\n",
    "        if s[i] == s[i-1]:\n",
    "            z[1] += 1\n",
    "        else:\n",
    "            break\n",
    "    \n",
    "    r, l = 0, 0\n",
    "    if z[1] > 0:\n",
    "        r, l = z[1], 1\n",
    "    \n",
    "    for k in range(2, len(s)):\n",
    "        assert z[k] == 0\n",
    "        if k > r:\n",
    "            # Case 1\n",
    "            for i in range(k, len(s)):\n",
    "                if s[i] == s[i-k]:\n",
    "                    z[k] += 1\n",
    "                else:\n",
    "                    break\n",
    "            r, l = k + z[k] - 1, k\n",
    "        else:\n",
    "            # Case 2\n",
    "            # Calculate length of beta\n",
    "            nbeta = r - k + 1\n",
    "            zkp = z[k - l]\n",
    "            if nbeta > zkp:\n",
    "                # Case 2a: Zkp wins\n",
    "                z[k] = zkp\n",
    "            else:\n",
    "                # Case 2b: Compare characters just past r\n",
    "                nmatch = 0\n",
    "                for i in range(r+1, len(s)):\n",
    "                    if s[i] == s[i - k]:\n",
    "                        nmatch += 1\n",
    "                    else:\n",
    "                        break\n",
    "                l, r = k, r + nmatch\n",
    "                z[k] = r - k + 1\n",
    "    return z\n",
    "\n",
    "\n",
    "def n_array(s):\n",
    "    \"\"\" Compile the N array (Gusfield theorem 2.2.2) from the Z array \"\"\"\n",
    "    return z_array(s[::-1])[::-1]\n",
    "\n",
    "\n",
    "def big_l_prime_array(p, n):\n",
    "    \"\"\" Compile L' array (Gusfield theorem 2.2.2) using p and N array.\n",
    "        L'[i] = largest index j less than n such that N[j] = |P[i:]| \"\"\"\n",
    "    lp = [0] * len(p)\n",
    "    for j in range(len(p)-1):\n",
    "        i = len(p) - n[j]\n",
    "        if i < len(p):\n",
    "            lp[i] = j + 1\n",
    "\n",
    "    return lp\n",
    "\n",
    "\n",
    "def big_l_array(p, lp):\n",
    "    \"\"\" Compile L array (Gusfield theorem 2.2.2) using p and L' array.\n",
    "        L[i] = largest index j less than n such that N[j] >= |P[i:]| \"\"\"\n",
    "    l = [0] * len(p)\n",
    "    l[1] = lp[1]\n",
    "    for i in range(2, len(p)):\n",
    "        l[i] = max(l[i-1], lp[i])\n",
    "    return l\n",
    "\n",
    "\n",
    "def small_l_prime_array(n):\n",
    "    \"\"\" Compile lp' array (Gusfield theorem 2.2.4) using N array. \"\"\"\n",
    "    small_lp = [0] * len(n)\n",
    "    for i in range(len(n)):\n",
    "        if n[i] == i+1:  # prefix matching a suffix\n",
    "            small_lp[len(n)-i-1] = i+1\n",
    "    for i in range(len(n)-2, -1, -1):  # \"smear\" them out to the left\n",
    "        if small_lp[i] == 0:\n",
    "            small_lp[i] = small_lp[i+1]\n",
    "    return small_lp\n",
    "\n",
    "\n",
    "def good_suffix_table(p):\n",
    "    \"\"\" Return tables needed to apply good suffix rule. \"\"\"\n",
    "    n = n_array(p)\n",
    "    lp = big_l_prime_array(p, n)\n",
    "    return lp, big_l_array(p, lp), small_l_prime_array(n)\n",
    "\n",
    "\n",
    "def good_suffix_mismatch(i, big_l_prime, small_l_prime):\n",
    "    \"\"\" Given a mismatch at offset i, and given L/L' and l' arrays,\n",
    "        return amount to shift as determined by good suffix rule. \"\"\"\n",
    "    length = len(big_l_prime)\n",
    "    assert i < length\n",
    "    if i == length - 1:\n",
    "        return 0\n",
    "    i += 1  # i points to leftmost matching position of P\n",
    "    if big_l_prime[i] > 0:\n",
    "        return length - big_l_prime[i]\n",
    "    return length - small_l_prime[i]\n",
    "\n",
    "\n",
    "def good_suffix_match(small_l_prime):\n",
    "    \"\"\" Given a full match of P to T, return amount to shift as\n",
    "        determined by good suffix rule. \"\"\"\n",
    "    return len(small_l_prime) - small_l_prime[1]\n",
    "\n",
    "\n",
    "def dense_bad_char_tab(p, amap):\n",
    "    \"\"\" Given pattern string and list with ordered alphabet characters, create\n",
    "        and return a dense bad character table.  Table is indexed by offset\n",
    "        then by character. \"\"\"\n",
    "    tab = []\n",
    "    nxt = [0] * len(amap)\n",
    "    for i in range(0, len(p)):\n",
    "        c = p[i]\n",
    "        assert c in amap\n",
    "        tab.append(nxt[:])\n",
    "        nxt[amap[c]] = i+1\n",
    "    return tab\n",
    "\n",
    "\n",
    "class BoyerMoore(object):\n",
    "    \"\"\" Encapsulates pattern and associated Boyer-Moore preprocessing. \"\"\"\n",
    "    \n",
    "    def __init__(self, p, alphabet='ACGT'):\n",
    "        self.p = p\n",
    "        self.alphabet = alphabet\n",
    "\n",
    "        # Create map from alphabet characters to integers\n",
    "        self.amap = {}\n",
    "        for i in range(len(self.alphabet)):\n",
    "            self.amap[self.alphabet[i]] = i\n",
    "\n",
    "        # Make bad character rule table\n",
    "        self.bad_char = dense_bad_char_tab(p, self.amap)\n",
    "\n",
    "        # Create good suffix rule table\n",
    "        _, self.big_l, self.small_l_prime = good_suffix_table(p)\n",
    "\n",
    "    def bad_character_rule(self, i, c):\n",
    "        \"\"\" Return # skips given by bad character rule at offset i \"\"\"\n",
    "        assert c in self.amap\n",
    "        ci = self.amap[c]\n",
    "        assert i > (self.bad_char[i][ci]-1)\n",
    "        return i - (self.bad_char[i][ci]-1)\n",
    "\n",
    "    def good_suffix_rule(self, i):\n",
    "        \"\"\" Given a mismatch at offset i, return amount to shift\n",
    "            as determined by (weak) good suffix rule. \"\"\"\n",
    "        length = len(self.big_l)\n",
    "        assert i < length\n",
    "        if i == length - 1:\n",
    "            return 0\n",
    "        i += 1  # i points to leftmost matching position of P\n",
    "        if self.big_l[i] > 0:\n",
    "            return length - self.big_l[i]\n",
    "        return length - self.small_l_prime[i]\n",
    "\n",
    "    def match_skip(self):\n",
    "        \"\"\" Return amount to shift in case where P matches T \"\"\"\n",
    "        return len(self.small_l_prime) - self.small_l_prime[1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e6db67e6-d108-4b9a-86e4-e83a86834ba3",
   "metadata": {},
   "outputs": [],
   "source": [
    "def boyer_moore(p, p_bm, t):\n",
    "    \"\"\" Do Boyer-Moore matching \"\"\"\n",
    "    i = 0\n",
    "    occurrences = []\n",
    "    align_count = 0 #number of alignment when matching the string with the reference genome\n",
    "    while i < len(t) - len(p) + 1:\n",
    "        shift = 1\n",
    "        mismatched = False\n",
    "        align_count += 1\n",
    "        for j in range(len(p)-1, -1, -1):\n",
    "            if p[j] != t[i+j]:\n",
    "                skip_bc = p_bm.bad_character_rule(j, t[i+j])\n",
    "                skip_gs = p_bm.good_suffix_rule(j)\n",
    "                shift = max(shift, skip_bc, skip_gs)\n",
    "                mismatched = True\n",
    "                break\n",
    "        if not mismatched:\n",
    "            occurrences.append(i)\n",
    "            skip_gs = p_bm.match_skip()\n",
    "            shift = max(shift, skip_gs)\n",
    "        i += shift\n",
    "    return occurrences,align_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "cd6d9aa2-62b9-4e1c-8d17-7261bd2bc79f",
   "metadata": {},
   "outputs": [],
   "source": [
    "s_bm = BoyerMoore(s, alphabet='ACGT')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "cffc23fc-440c-48cc-8daa-75a59def9080",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([56922], 127974)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "boyer_moore(s, s_bm, genome)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "cbf9aab8-91bc-4556-9643-9ed3ce64e723",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Reading genome from -f...\n",
      "Genome length: 800000 bp\n",
      "Building 8-mer index...\n",
      "Index built successfully\n",
      "Finding approximate matches (up to 2 mismatches, forward strand only)...\n",
      "\n",
      "Results for forward strand only:\n",
      "The Alu-derived pattern occurs 19 times with up to 2 mismatches\n",
      "Total index hits during search: 90\n",
      "\n",
      "First 5 match positions: [56922, 84641, 147558, 160162, 160729]\n",
      "\n",
      "Sample matches (position, sequence, mismatches):\n",
      "Position 56922: GGCGCGGTGGCTCACGCCTGTAAT, Mismatches: 0\n",
      "Position 84641: GGCGCGGTGGCTCATGCCTGTAAT, Mismatches: 1\n",
      "Position 147558: GGCGCGGTGGCTCATGCCTGTAAT, Mismatches: 1\n"
     ]
    }
   ],
   "source": [
    "import bisect\n",
    "import sys\n",
    "\n",
    "class Index(object):\n",
    "    def __init__(self, t, k):\n",
    "        ''' Create index from all substrings of size 'length' '''\n",
    "        self.k = k  # k-mer length (k)\n",
    "        self.index = []\n",
    "        for i in range(len(t) - k + 1):  # for each k-mer\n",
    "            self.index.append((t[i:i+k], i))  # add (k-mer, offset) pair\n",
    "        self.index.sort()  # alphabetize by k-mer\n",
    "    \n",
    "    def query(self, p):\n",
    "        ''' Return index hits for first k-mer of P '''\n",
    "        kmer = p[:self.k]  # query with first k-mer\n",
    "        i = bisect.bisect_left(self.index, (kmer, -1))  # binary search\n",
    "        hits = []\n",
    "        while i < len(self.index):  # collect matching index entries\n",
    "            if self.index[i][0] != kmer:\n",
    "                break\n",
    "            hits.append(self.index[i][1])\n",
    "            i += 1\n",
    "        return hits\n",
    "\n",
    "def read_genome(filename):\n",
    "    \"\"\"Read a genome from a FASTA or text file\"\"\"\n",
    "    genome = ''\n",
    "    with open(r\"C:\\Users\\liamm\\Desktop\\chr1.GRCh38.excerpt.fasta\", 'r') as f:\n",
    "        for line in f:\n",
    "            if not line.startswith('>'):  # skip header lines\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "\n",
    "def reverse_complement(s):\n",
    "    \"\"\"Return the reverse complement of a DNA string\"\"\"\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N': 'N'}\n",
    "    return ''.join(complement.get(base, base) for base in reversed(s))\n",
    "\n",
    "def approximate_match_index_with_hit_count(p, t, index, n=2, include_reverse=False):\n",
    "   \n",
    "    segment_length = index.k  # 8-mer length\n",
    "    all_matches = set()  # Use a set to avoid counting matches multiple times\n",
    "    total_hits = 0  # Counter for total index hits\n",
    "    match_details = {\n",
    "        'forward': {},  \n",
    "        'reverse': {}   \n",
    "    }\n",
    "    \n",
    "    patterns = [(p, 'forward')]\n",
    "    if include_reverse:\n",
    "        patterns.append((reverse_complement(p), 'reverse'))\n",
    "    \n",
    "    for pattern, direction in patterns:\n",
    "        # Divide the pattern into n+1 segments to guarantee at least one exact match\n",
    "        for i in range(n + 1):\n",
    "            start = i * segment_length\n",
    "            end = min((i + 1) * segment_length, len(pattern))\n",
    "            \n",
    "            # Skip if segment would be too short (edge case)\n",
    "            if end - start < segment_length:\n",
    "                continue\n",
    "            \n",
    "            # Use index to find exact matches for this segment\n",
    "            segment = pattern[start:end]\n",
    "            hits = index.query(segment)\n",
    "            \n",
    "            # Count total index hits for this segment\n",
    "            total_hits += len(hits)\n",
    "            \n",
    "            # Check each potential match\n",
    "            for hit in hits:\n",
    "                # Adjust hit position to account for segment offset\n",
    "                hit_adjusted = hit - start\n",
    "                \n",
    "                # Skip if adjusted position would be out of bounds\n",
    "                if hit_adjusted < 0 or hit_adjusted + len(pattern) > len(t):\n",
    "                    continue\n",
    "                \n",
    "                # Count mismatches between pattern and text at adjusted position\n",
    "                mismatches = 0\n",
    "                for j in range(len(pattern)):\n",
    "                    if hit_adjusted + j < len(t) and pattern[j] != t[hit_adjusted + j]:\n",
    "                        mismatches += 1\n",
    "                        if mismatches > n:\n",
    "                            break\n",
    "                \n",
    "                # Add position to matches if within mismatch threshold\n",
    "                if mismatches <= n:\n",
    "                    all_matches.add(hit_adjusted)\n",
    "                    match_details[direction][hit_adjusted] = mismatches\n",
    "    \n",
    "    return sorted(list(all_matches)), total_hits, match_details\n",
    "\n",
    "def main():\n",
    "    \"\"\"Main function to run the approximate matching algorithm\"\"\"\n",
    "    # Alu-derived pattern\n",
    "    p = \"GGCGCGGTGGCTCACGCCTGTAAT\"\n",
    "    \n",
    "    # Check if a filename is provided as a command-line argument\n",
    "    if len(sys.argv) > 1:\n",
    "        filename = sys.argv[1]\n",
    "    else:\n",
    "        filename = input(\"Enter the path to the chromosome file: \")\n",
    "    \n",
    "    # Read the chromosome sequence\n",
    "    print(f\"Reading genome from {filename}...\")\n",
    "    t = read_genome(filename)\n",
    "    print(f\"Genome length: {len(t)} bp\")\n",
    "    \n",
    "    # Build the index\n",
    "    print(\"Building 8-mer index...\")\n",
    "    index = Index(t, 8)\n",
    "    print(\"Index built successfully\")\n",
    "    \n",
    "    # Find approximate matches without reverse complements (as specified in the problem)\n",
    "    print(\"Finding approximate matches (up to 2 mismatches, forward strand only)...\")\n",
    "    matches, total_hits, match_details = approximate_match_index_with_hit_count(p, t, index, n=2, include_reverse=False)\n",
    "    \n",
    "    print(f\"\\nResults for forward strand only:\")\n",
    "    print(f\"The Alu-derived pattern occurs {len(matches)} times with up to 2 mismatches\")\n",
    "    print(f\"Total index hits during search: {total_hits}\")\n",
    "    \n",
    "    # If you want to include reverse complements, uncomment the following:\n",
    "    \"\"\"\n",
    "    print(\"\\nNow searching with reverse complements included...\")\n",
    "    matches_with_rc, total_hits_with_rc, match_details_with_rc = approximate_match_index_with_hit_count(\n",
    "        p, t, index, n=2, include_reverse=True)\n",
    "    \n",
    "    forward_matches = sum(1 for pos in match_details_with_rc['forward'])\n",
    "    reverse_matches = sum(1 for pos in match_details_with_rc['reverse'])\n",
    "    \n",
    "    print(f\"\\nResults including reverse complements:\")\n",
    "    print(f\"The Alu-derived pattern occurs {len(matches_with_rc)} times total with up to 2 mismatches\")\n",
    "    print(f\"Forward strand matches: {forward_matches}\")\n",
    "    print(f\"Reverse strand matches: {reverse_matches}\")\n",
    "    print(f\"Total index hits during search: {total_hits_with_rc}\")\n",
    "    \"\"\"\n",
    "    \n",
    "    # Print the first few matches for verification\n",
    "    if matches:\n",
    "        print(\"\\nFirst 5 match positions:\", matches[:5])\n",
    "        print(\"\\nSample matches (position, sequence, mismatches):\")\n",
    "        count = 0\n",
    "        for pos in sorted(match_details['forward'].keys()):\n",
    "            if count >= 3:\n",
    "                break\n",
    "            match_seq = t[pos:pos+len(p)]\n",
    "            mismatches = match_details['forward'][pos]\n",
    "            print(f\"Position {pos}: {match_seq}, Mismatches: {mismatches}\")\n",
    "            count += 1\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "961b3fe6-da47-4b7d-a96f-61672ba939d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Approximate matches (up to 2 mismatches) for 24-mer: [56922, 84641, 147558, 191452, 262042, 273669, 364263, 465647, 635931, 657496, 681737, 717706, 747359]\n",
      "Total number of index hits for 24-mer: 35\n"
     ]
    }
   ],
   "source": [
    "import bisect\n",
    "\n",
    "class SubseqIndex(object):\n",
    "    \"\"\" Holds a subsequence index for a text T \"\"\"\n",
    "\n",
    "    def __init__(self, t, k, ival):\n",
    "        \"\"\"\n",
    "        Initializes a SubseqIndex object.\n",
    "\n",
    "        Args:\n",
    "            t (str): The text (genome) to be indexed.\n",
    "            k (int): The k-mer length for the index.\n",
    "            ival (int): The interval between indexed k-mers.\n",
    "        \"\"\"\n",
    "        self.k = k  # num characters per subsequence extracted\n",
    "        self.ival = ival  # space between them; 1=adjacent, 2=every other, etc\n",
    "        self.index = []\n",
    "        self.span = 1 + ival * (k - 1)\n",
    "        for i in range(len(t) - self.span + 1):  # for each subseq\n",
    "            self.index.append((t[i:i+self.span:ival], i))  # add (subseq, offset)\n",
    "        self.index.sort()  # alphabetize by subseq\n",
    "\n",
    "    def query(self, p):\n",
    "        \"\"\"\n",
    "        Return index hits for first subseq of p.\n",
    "\n",
    "        Args:\n",
    "            p (str): The pattern to query.\n",
    "\n",
    "        Returns:\n",
    "            list: A list of indices in t where the k-mer of p occurs.\n",
    "        \"\"\"\n",
    "        subseq = p[:self.span:self.ival]  # query with first subseq\n",
    "        i = bisect.bisect_left(self.index, (subseq, -1))  # binary search\n",
    "        hits = []\n",
    "        while i < len(self.index):  # collect matching index entries\n",
    "            if self.index[i][0] != subseq:\n",
    "                break\n",
    "            hits.append(self.index[i][1])\n",
    "            i += 1\n",
    "        return hits\n",
    "\n",
    "\n",
    "def find_approximate_matches_24mer(pattern, text, index, max_mismatches=2):\n",
    "    \"\"\"\n",
    "    Finds approximate matches of a 24-length pattern in a text using a SubseqIndex\n",
    "    with k=8 and ival=3, allowing up to 2 mismatches.\n",
    "\n",
    "    Args:\n",
    "        pattern (str): The pattern to search for (length 24).\n",
    "        text (str): The text (genome) to search in.\n",
    "        index (SubseqIndex): A SubseqIndex object built on the text.\n",
    "        max_mismatches (int, optional): The maximum number of allowed mismatches. Defaults to 2.\n",
    "\n",
    "    Returns:\n",
    "        tuple: A tuple containing:\n",
    "            - list: A list of indices in text where approximate matches of the pattern occur.\n",
    "            - int: The total number of index hits.\n",
    "    \"\"\"\n",
    "    if len(pattern) != 24:\n",
    "        raise ValueError(\"Pattern length must be 24\")\n",
    "    if max_mismatches > 2:\n",
    "        raise ValueError(\"Maximum mismatches allowed is 2\")\n",
    "\n",
    "    hits = index.query(pattern)\n",
    "    total_index_hits = len(hits)  # Count the total number of index hits\n",
    "    matches = []\n",
    "\n",
    "    for hit in hits:\n",
    "        mismatches = 0\n",
    "        for i in range(len(pattern)):\n",
    "            if text[hit + i] != pattern[i]:\n",
    "                mismatches += 1\n",
    "                if mismatches > max_mismatches:\n",
    "                    break\n",
    "        if mismatches <= max_mismatches:\n",
    "            matches.append(hit)\n",
    "\n",
    "    return sorted(list(set(matches))), total_index_hits  # Remove duplicates and sort, return total hits\n",
    "\n",
    "\n",
    "\n",
    "def readGenome(filename):\n",
    "    genome = ''\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip()\n",
    "    return genome\n",
    "\n",
    "if __name__ == '__main__':\n",
    "    # Example usage:\n",
    "    filename = r\"C:\\Users\\liamm\\Desktop\\chr1.GRCh38.excerpt.fasta\"  \n",
    "    genome = readGenome(filename)\n",
    "    pattern_24mer = \"GGCGCGGTGGCTCACGCCTGTAAT\"  \n",
    "    k = 8\n",
    "    ival = 3\n",
    "    index = SubseqIndex(genome, k, ival)\n",
    "    approximate_matches_24mer, total_hits_24mer = find_approximate_matches_24mer(pattern_24mer, genome, index)\n",
    "    print(f\"Approximate matches (up to 2 mismatches) for 24-mer: {approximate_matches_24mer}\")\n",
    "    print(f\"Total number of index hits for 24-mer: {total_hits_24mer}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "983d32e7-1e43-43e5-ae8d-887a807587f5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:torch_env]",
   "language": "python",
   "name": "conda-env-torch_env-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
